src/ferny/transport.py

```diff
@@ -0,0 +1,371 @@
+from .ssh_errors import get_exception_for_ssh_stderr
+
+logger = logging.getLogger(__name__)
+
+P = TypeVar('P', bound=asyncio.Protocol)
+
+
+class FernyTransport(asyncio.Transport, asyncio.SubprocessProtocol):
+    # These get initialized in connection_made(), not __init__(), but we don't
+    # list them as Optional because we won't return the object to the user
+    # until after the initialization is complete.
+    _subprocess_transport: 'asyncio.SubprocessTransport | None' = None
+    _stdin_transport: 'asyncio.WriteTransport | None' = None
+    _stdout_transport: 'asyncio.ReadTransport | None' = None
+    _agent: InteractionAgent
+    _is_ssh: bool
+
+    # We record events that might build towards a connection termination here
+    # and consider them from _consider_disconnect() in order to try to get the
+    # best possible Exception for the protocol, rather than just taking the
+    # first one (which is likely to be somewhat random).
+    _lost_connections: 'dict[int | None, Exception | None]'
+    _stderr_output: 'str | None' = None
+    _returncode: 'int | None' = None
+
+    # Transient state
+    _exec_task: 'asyncio.Task[tuple[asyncio.SubprocessTransport, FernyTransport]]'
+    _disconnected: bool = False
+
+    # This one comes in via __init__.
+    _protocol: asyncio.Protocol
+
+    @classmethod
+    def spawn(
+        cls: 'type[typing.Self]',
+        protocol_factory: Callable[[], P],
+        args: Sequence[str],
+        loop: 'asyncio.AbstractEventLoop | None' = None,
+        interaction_handlers: Sequence[InteractionHandler] = (),
+        is_ssh: bool = True,
+        **kwargs: Any
+    ) -> 'tuple[typing.Self, P]':
+        """Connects a FernyTransport to a protocol, using the given command.
+
+        This spawns the command and connects the stdin and stdout of the
+        command to the protocol returned by the factory.  The stderr is handled
+        via ferny's InteractionHandler.
+
+        If spawning the process fails then connection_lost() will be
+        immediately called with the relevant OSError, even before
+        connection_made() was called.  This is somewhat non-standard behaviour,
+        but is the easiest way to report these errors without making this
+        function async.
+
+        Once the process is successfully executed, connection_made() will be
+        called and the transport can be used as normal.  connection_lost() will
+        be called if the process exits or another error occurs.
+
+        The return value of this function is the transport, but it exists in a
+        semi-initialized state.  You can call .close() on it, but nothing else.
+        Once .connection_made() is called, you can call all the other
+        functions.
+
+        The return value of this function must be held alive until at least one
+        of these things happens:
+          - it has .close() called on it
+          - .connection_lost() is called on your Protocol
```

Comment:
The example awaits the `.done` future, this could be added here?

Reply:
This is actually no longer accurate.  The condition for dropping the transport is now, unconditionally, that `connection_lost()` is called.  I'll fix that up.

Reply:
> The example awaits the `.done` future, this could be added here?

The `.done` future is implemented inside of the example itself, from its `connection_lost()` handler:

```python
    def connection_lost(self, exc: Exception | None) -> None:
        if exc is not None:
            self.done.set_exception(exc)
        else:
            self.done.set_result(None)
```

So it's effectively doing exactly this.